k4GaudiPandora/src/DDCaloDigi.cc

```diff
@@ -1061,674 +905,333 @@ float DDCaloDigi::analogueHcalCalibCoeff(CHT::Layout caloLayout, int layer) {
       break;
     }
   }
-
   return calib_coeff;
 }
 
-float DDCaloDigi::analogueEcalCalibCoeff(int layer) {
+float DDCaloDigi::analogueEcalCalibCoeff(int layer) const {
   float calib_coeff = 0;
 
   // retrieve calibration constants
-  for (unsigned int k(0); k < _ecalLayers.size(); ++k) {
+  for (unsigned int k(0); k < m_ecalLayers.size(); ++k) {
     int min, max;
     if (k == 0) {
       min = 0;
     } else {
-      min = _ecalLayers[k - 1];
+      min = m_ecalLayers[k - 1];
     }
-    max = _ecalLayers[k];
+    max = m_ecalLayers[k];
     if (layer >= min && layer < max) {
-      calib_coeff = _calibrCoeffEcal[k];
+      calib_coeff = m_calibrCoeffEcal[k];
       break;
     }
   }
-
   return calib_coeff;
 }
 
-float DDCaloDigi::ecalEnergyDigi(float energy, int id) {
+float DDCaloDigi::ecalEnergyDigi(float energy, int id, float event_correl_miscalib_ecal,
+                                 CLHEP::MTwistEngine& randomEngine) const {
   // some extra digi effects (daniel)
-  // controlled by _applyEcalDigi = 0 (none), 1 (silicon), 2 (scintillator)
+  // controlled by m_applyEcalDigi = 0 (none), 1 (silicon), 2 (scintillator)
 
   // small update for time-constant uncorrelated miscalibrations. DJ, Jan 2015
 
-  float e_out(energy);
-  if (_applyEcalDigi == 1)
-    e_out = siliconDigi(energy);  // silicon digi
-  else if (_applyEcalDigi == 2)
-    e_out = scintillatorDigi(energy, true);  // scintillator digi
-
+  float e_out = energy;
+  if (m_applyEcalDigi == 1) {
+    e_out = siliconDigi(energy, randomEngine);  // silicon digi
+  } else if (m_applyEcalDigi == 2) {
+    e_out = scintillatorDigi(energy, true, randomEngine);  // scintillator digi
+  }
   // add electronics dynamic range
   // Sept 2015: Daniel moved this to the ScintillatorDigi part, so it is applied before unfolding of sipm response
-  // if (_ecalMaxDynMip>0) e_out = min (e_out, _ecalMaxDynMip*_calibEcalMip);
+  // if (_ecalMaxDynMip>0) e_out = min (e_out, _ecalMaxDynMip*_calibEcalMip); ???
 
   // random miscalib
-  if (_misCalibEcal_uncorrel > 0) {
+  if (m_misCalibEcal_uncorrel > 0) {
     float miscal(0);
-    if (_misCalibEcal_uncorrel_keep) {
-      int id;
-      if (_ECAL_cell_miscalibs.find(id) !=
-          _ECAL_cell_miscalibs.end()) {  // this cell was previously seen, and a miscalib stored
-        miscal = _ECAL_cell_miscalibs[id];
+    if (m_misCalibEcal_uncorrel_keep) {
+      if (m_ECAL_cell_miscalibs.find(id) !=
+          m_ECAL_cell_miscalibs.end()) {  // this cell was previously seen, and a miscalib stored
+        miscal = m_ECAL_cell_miscalibs.at(id);
       } else {  // we haven't seen this one yet, get a miscalib for it
-        miscal                   = CLHEP::RandGauss::shoot(1.0, _misCalibEcal_uncorrel);
-        _ECAL_cell_miscalibs[id] = miscal;
+        miscal = CLHEP::RandGauss::shoot(&randomEngine, 1.0, m_misCalibEcal_uncorrel);
+        // FIXME: this is storing miscalibration globally for a run???
```

Comment:
I don't know how to properly implement this. We would need a global map of dead cells, or miscalibration parameters.

Reply:
This will be implemented later.